### ISSUE TITLE
"no matching cipher in this list found (pls report this)" for non-SNI only on LAN

### ISSUE THREAD

**Issue by welwood08** [drwetter/testssl.sh#446]
I was testing a service on my local network when I noticed an inconsistency in results between host name and IP address URIs. When I use the IP address, I get `Has server cipher order?     no matching cipher in this list found (pls report this)` but when I use the internal host name I get `Has server cipher order?     yes (OK)`.

After trying to bisect commits to pinpoint a regression range, I found instead that the problem was present at 8c044587c5ee2b131a94d49279c20477c8dd9031 (the initial FreeBSD build of the bundled OpenSSL from last year). The issue does not appear when using either the latest version of OpenSSL in FreeBSD ports ("1.0.2h  3 May 2016") or my jail's system OpenSSL ("1.0.1s-freebsd  1 Mar 2016") or my host's system OpenSSL ("1.0.1t-freebsd  3 May 2016").

Full debug logs using latest FreeBSD-built OpenSSL package, showing no error: [debug-openssl-bsdports.zip](https://github.com/drwetter/testssl.sh/files/417027/debug-openssl-bsdports.zip)
Full debug logs using latest testssl.sh-bundled OpenSSL, showing failure: [debug-openssl-bundled.zip](https://github.com/drwetter/testssl.sh/files/417028/debug-openssl-bundled.zip)

As of 424cf233d1a889e89471cc180592ec90f60e7af0 (latest at time of writing), the command that fails is [`$OPENSSL s_client $STARTTLS -cipher $list_fwd $BUGS -connect $NODEIP:$PORT $PROXY $SNI </dev/null 2>$ERRFILE >$TMPFILE`](https://github.com/drwetter/testssl.sh/blob/424cf233d1a889e89471cc180592ec90f60e7af0/testssl.sh#L3322), which on my system expands to `./bin/openssl.FreeBSD.amd64 s_client  -cipher DES-CBC3-SHA:RC4-MD5:DES-CBC-SHA:RC4-SHA:AES128-SHA:AES128-SHA256:AES256-SHA:ECDHE-RSA-AES128-SHA:DHE-RSA-AES128-SHA:DHE-RSA-AES256-SHA:ECDH-RSA-DES-CBC3-SHA:ECDH-RSA-AES128-SHA:ECDH-RSA-AES256-SHA:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-SHA256:DHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:DHE-DSS-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256:AES256-SHA256  -connect 10.10.10.10:443   </dev/null`.

I tested several servers by IP address using the latest bundled OpenSSL, with the following results:
- [ ] :x: Local, lighttpd/1.4.35 → `10441828:error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure:s23_clnt.c:769:`
- [ ] :x: Local, Mono-HTTPAPI/1.1 → `10441828:error:1407741A:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert decode error:s23_clnt.c:769:`
- [ ] :x: Local, unidentified version of Go's net/http → `10441828:error:140770FC:SSL routines:SSL23_GET_SERVER_HELLO:unknown protocol:s23_clnt.c:794:`
- [ ] :x: Local, nginx/1.11.3 (built with OpenSSL 1.0.2h  3 May 2016) → `10441828:error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure:s23_clnt.c:769:`
- [x] :ok: Internet, nginx/1.11.3 (built with OpenSSL 1.0.2h  3 May 2016)
- [x] :ok: Internet, Apache/2.4.7
- [x] :ok: Internet, lighttpd/2.0.0
- [x] :ok: Internet, news.newsguy.com:563 (#439)

A server cipher order check emitting `no matching cipher in this list found (pls report this)` is denoted by :x:, a result of either `yes (OK)` or `nope (NOT ok)` is denoted by :ok:.

Experimenting manually with OpenSSL and taking hints from the errors, I note that I receive the same errors from the local servers if I use the hostname and omit SNI (ie `-connect host.example.lan:443` without `-servername host.example.lan`). I was also able to make the errors go away by adding `-no_ssl2`. Presumably only the bundled OpenSSL is affected because it is deliberately built to support SSLv2? I am at a loss to explain why only servers on my local network are affected.

Please let me know if I've left out any pertinent information or of any additional tests I can perform (I control both tested nginx instances in case that's useful).

**Comment by drwetter** [comment]
Hi Will,

first: thx.

does openssl ciphers -V give you any additional information about the missing ciphers?

Specifying an IP address only targets the default host.

Cheers, Dirk

**Comment by welwood08** [comment]
There are no missing ciphers. `diff <(/usr/local/bin/openssl ciphers -V) <(./bin/openssl.FreeBSD.amd64 ciphers -V)` shows that the bundled version adds 26 more ciphers and subtracts none.

I know IP address requests would lead to the default host, however the server for which I first noticed this problem (local lighttpd) only has a default host and always serves both the same certificate and content.

Perhaps I was too verbose initially and the real problem will be clearer with a focused test on just that local lighttpd instance:

```sh
#!/usr/bin/env bash

list_fwd='DES-CBC3-SHA:RC4-MD5:DES-CBC-SHA:RC4-SHA:AES128-SHA:AES128-SHA256:AES256-SHA:ECDHE-RSA-AES128-SHA:DHE-RSA-AES128-SHA:DHE-RSA-AES256-SHA:ECDH-RSA-DES-CBC3-SHA:ECDH-RSA-AES128-SHA:ECDH-RSA-AES256-SHA:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES128-SHA256:DHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-AES256-SHA384:ECDHE-RSA-AES256-SHA:DHE-DSS-AES256-GCM-SHA384:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-GCM-SHA256:AES256-SHA256'
STARTTLS=''
BUGS=''
PROXY=''
PORT=443

# FreeBSD ports vs bundled:
for OPENSSL in '/usr/local/bin/openssl' './bin/openssl.FreeBSD.amd64'; do
  # IP vs hostname which resolves to that same IP:
  for NODEIP in '10.10.10.10' 'example.lan'; do
    # non-SNI vs any random SNI:
    for SNI in '' "-servername $(openssl rand -hex 16)"; do
      # testssl.sh line 3322 (minus output redirection)
      $OPENSSL s_client $STARTTLS -cipher $list_fwd $BUGS -connect $NODEIP:$PORT $PROXY $SNI </dev/null
    done
  done
done
```

OPENSSL= | NODEIP= | SNI= | Result
----------- | ---------- | ----- | ------
'/usr/local/bin/openssl' | '10.10.10.10' | '' | :ok:
'/usr/local/bin/openssl' | '10.10.10.10' | "-servername $(openssl rand -hex 16)" | :ok:
'/usr/local/bin/openssl' | 'example.lan' | '' | :ok:
'/usr/local/bin/openssl' | 'example.lan' | "-servername $(openssl rand -hex 16)" | :ok:
'./bin/openssl.FreeBSD.amd64' | '10.10.10.10' | '' | :x:
'./bin/openssl.FreeBSD.amd64' | '10.10.10.10' | "-servername $(openssl rand -hex 16)" | :ok:
'./bin/openssl.FreeBSD.amd64' | 'example.lan' | '' | :x:
'./bin/openssl.FreeBSD.amd64' | 'example.lan' | "-servername $(openssl rand -hex 16)" | :ok:

In all the :ok: cases, the server returns exactly the same certificate and the `Master-Key:` line has a long hex string (this seems to be what testssl.sh is expecting).
In both the :x: cases, the full output looks like this:
```
CONNECTED(00000004)
10441828:error:14077410:SSL routines:SSL23_GET_SERVER_HELLO:sslv3 alert handshake failure:s23_clnt.c:769:
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 7 bytes and written 88 bytes
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : 0000
    Session-ID:
    Session-ID-ctx:
    Master-Key:
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1471270770
    Timeout   : 300 (sec)
    Verify return code: 0 (ok)
---
```

I get similar results for all local target servers (with varying SSL23_GET_SERVER_HELLO errors as shown in the 1st post). I still have not been able to reproduce for any target servers outside my local network.

**Comment by drwetter** [comment]
Sorry, it's too hard to track this down without having access, especially when it probably boils down to problems with openssl version. Some parts are also confusing, i.e. the random servername part. This seems to have nothing to do with testssl.sh

> In both the :x: cases, the full output looks like this:

That reads: no connection

> After trying to bisect commits to pinpoint a regression range, I found instead that the problem was 
> present at 8c04458 (the initial FreeBSD build of the bundled OpenSSL from last year). The issue does > not appear when using either the latest version of OpenSSL in FreeBSD ports ("1.0.2h 3 May 2016") or > my jail's system OpenSSL ("1.0.1s-freebsd 1 Mar 2016") or my host's system OpenSSL ("1.0.1t-
> freebsd 3 May 2016").

Why don't you report what happens with the recent FreeBSD binary?

**Comment by welwood08** [comment]
Hi Dirk, thanks for looking at this again.
> Some parts are also confusing, i.e. the random servername part.

I was just using that as an example to demonstrate that it doesn't matter what server name I provide for SNI, the results are consistent - any SNI results in success, omitting SNI results in failure.
> That reads: no connection

Except it has read 7 bytes? This is the root of my confusion. As you can see, I have scripted the test in a tight nested loop so it's unlikely to be losing connection consistently for only the 5th and 7th of the 8 test cases.
> Why don't you report what happens with the recent FreeBSD binary?

```
$> ./bin/openssl.FreeBSD.amd64 version -v -b
OpenSSL 1.0.2-chacha (1.0.2i-dev)
built on: Wed Jun 22 19:49:53 2016
```
Of course I've been doing that, it's what I was using when I first noticed a problem. That particular paragraph is just to say that I've also checked the earliest bundled version and found the same result.

With all that said, I've continued the investigation and discovered that my test script passes all 8 cases when directed at a non-jailed nginx instance on the FreeBSD host system (tests to jailed servers on the same machine fail in the same way as tests to other machines on the network). This now makes me suspect this could be at least partially a FreeBSD networking issue, possibly related to jails on the host using VIMAGE to share the hardware NIC. It doesn't explain why system binaries always pass the test script, so I still think there's some critical difference in the bundled fork (maybe the IPv6 patch has a bug? maybe one of the patches used to build the FreeBSD ports fixes a bug from upstream?).

In any case, since it now seems clear testssl.sh itself is not at fault directly and I'm not sure I want to make my local network available for external testing, I'm going to close this issue - sorry for the noise. Maybe some day I'll resume investigation by building PeterMosmans/openssl myself to see if that offers some clues, or maybe whenever that fork rebases onto 1.1 (with native IPv6 support) and the bundle is updated the problem goes away.

**Comment by drwetter** [comment]
Am 2. September 2016 04:11:02 MESZ, schrieb Will Elwood <notifications@github.com>:
>Hi Dirk, thanks for looking at this again.

np

>> That reads: no connection
>
>Except it has read 7 bytes? This is the root of my confusion. 

TCP: yes. TLS: no.

No cipher, no master key. And an alert.

> sorry for the noise.

No worries 

Cheers, Dirk


-- 
Sent from my mobile. Excuse my brevity&typos+the phone's autocorrection

**Comment by welwood08** [comment]
Hi, I updated my installation of testssl today and noticed this bug was no longer present so thought I'd better check my conclusion even though this is already closed.

To be sure the bug wasn't fixed by intervening system upgrades I undid my update of testssl to prove the bug was still there in the older version - it was.

After a bit of bisection, I found that cca1b4989091ee6ab646e58af80a0fc239617bdf was the fix. To confirm, the test script I posted above can be modified in a similar way (`sed "s/for SNI in ''/for SNI in '-no_ssl2'/"`). The 4 LAN servers I originally tested that failed now succeed with this 1 tiny change.

If I understand correctly, this issue was caused by the bundled version of openssl sending an SSLv2 hello when not specifying SNI (presumably this would happen for any openssl built with SSL2 enabled) and hence the servers I was testing against aborting the connection. This also seems to hold true outside my LAN, so I made a mistake in my original diagnosis there - I guess it was working only for servers that accepted SSLv2 hello.

I don't think from the commit message or inline comments that it was intended to fix this issue, nevertheless I no longer see the message that prompted me to create this issue - thanks!
